events/handler.go

```diff
@@ -0,0 +1,230 @@
+	return e[i].Timestamp.Before(e[j].Timestamp)
+}
+
+// sorts and returns up to the last MaxEventsReturned chronological elements
+func getMaxEventsReturned(request *Request, eSlice EventSlice) EventSlice {
+	sort.Sort(eSlice)
+	n := request.MaxEventsReturned
+	if n >= eSlice.Len() || n <= 0 {
+		return eSlice
+	}
+	return eSlice[eSlice.Len()-n:]
+}
+
+// If the request wants all subcontainers, this returns if the request's
+// container path is a prefix of the event container path.  Otherwise,
+// it checks that the container paths of the event and request are
+// equivalent
+func checkIfIsSubcontainer(request *Request, event *Event) bool {
+	if request.IncludeSubcontainers == true {
+		return strings.HasPrefix(event.ContainerName, request.ContainerName)
+	}
+	return event.ContainerName == request.ContainerName
+
+}
+
+// determines if an event occurs within the time set in the request object and is the right type
+func checkIfEventSatisfiesRequest(request *Request, event *Event) bool {
+	startTime := request.StartTime
+	endTime := request.EndTime
+	eventTime := event.Timestamp
+	if !startTime.IsZero() {
+		if startTime.After(eventTime) {
+			return false
+		}
+	}
+	if !endTime.IsZero() {
+		if endTime.Before(eventTime) {
+			return false
+		}
+	}
+	if request.EventType[event.EventType] != true {
+		return false
+	}
+	if request.ContainerName != "" {
+		return checkIfIsSubcontainer(request, event)
+	}
+	return true
+}
+
+// method of Events object that screens Event objects found in the eventlist
+// attribute and if they fit the parameters passed by the Request object,
+// adds it to a slice of *Event objects that is returned. If both MaxEventsReturned
+// and StartTime/EndTime are specified in the request object, then only
+// up to the most recent MaxEventsReturned events in that time range are returned.
+func (self *events) GetEvents(request *Request) (EventSlice, error) {
+	returnEventList := EventSlice{}
+	for _, e := range self.eventlist {
+		if checkIfEventSatisfiesRequest(request, e) {
+			returnEventList = append(returnEventList, e)
+		}
+	}
+	returnEventList = getMaxEventsReturned(request, returnEventList)
+	return returnEventList, nil
+}
+
+// method of Events object that maintains an *Event channel passed by the user.
+// When an event is added by AddEvents that satisfies the parameters in the passed
+// Request object it is fed to the channel. The StartTime and EndTime of the watch
+// request should be uninitialized because the purpose is to watch indefinitely
+// for events that will happen in the future
+func (self *events) WatchEvents(outChannel chan *Event, request *Request) error {
+	if !request.StartTime.IsZero() || !request.EndTime.IsZero() {
+		return errors.New(
+			"for a call to watch, request.StartTime and request.EndTime must be uninitialized")
+	}
+	newWatcher := newWatch(request, outChannel)
+	self.watchers = append(self.watchers, newWatcher)
+	for event := range newWatcher.channel {
```

Comment:
We can just return, the user will do the actual listening (and I think this will listen and send on the same channel no? Since newWatch() takes outChannel)